src/cocotb/regression.py

```diff
@@ -189,105 +230,49 @@ def discover_tests(
             self.log.debug("Searching for tests in module %s", module_name)
             self.log.debug("Python Path: %s", ",".join(sys.path))
             self.log.debug("PWD: %s", os.getcwd())
-            __import__(module_name)
+            mod = import_module(module_name)
+
+            if not hasattr(mod, "__cocotb_tests__"):
+                raise RuntimeError(f"No tests were discovered in module: {module_name}")
 
-        self.ntests = len(self._queue)
+            for test in mod.__cocotb_tests__:
+                self.register_test(test)
 
         # error if no tests were discovered
-        if self.ntests == 0:
+        if not self._test_queue:
             modules_str = ", ".join(repr(m) for m in modules)
             raise RuntimeError(f"No tests were discovered in modules: {modules_str}")
 
-        # if no filters, accept all tests
-        if not filters:
-            return
-
-        # else mark all tests as excluded
-        self.ntests = 0
-        for test in self._queue:
-            test.included = False
+    def filter_tests(self, filters: Sequence[str]) -> None:
+        included: List[Test] = []
+        excluded: List[Test] = []
 
         # include tests that match filter
-        for test in self._queue:
+        for test in self._test_queue:
             for filter in filters:
                 if re.search(filter, test.fullname):
-                    self.ntests += 1
-                    test.included = True
+                    included.append(test)
                     break
             else:
                 self.log.debug("Filtered out test %s", test.fullname)
+                excluded.append(test)
 
-        if self.ntests == 0:
+        if not included:
             self.log.warning(
                 "No tests left after filtering with: %s",
                 ", ".join(repr(f) for f in filters),
             )
 
-        self._queue.sort(key=lambda test: (test.stage, test._id))
+        self._test_queue = included
+        self._filtered_tests = excluded
 
-    def register_test(
-        self,
-        func: Callable[..., Coroutine[Any, Any, None]],
-        name: Optional[str] = None,
-        module: Optional[str] = None,
-        timeout_time: Optional[float] = None,
-        timeout_unit: str = "step",
-        expect_fail: bool = False,
-        expect_error: Union[Type[Exception], Sequence[Type[Exception]]] = (),
-        skip: bool = False,
-        stage: int = 0,
-    ) -> None:
+    def register_test(self, test: Test) -> None:
```

Comment:
May want to add description of `test` arg for docs, right now description is empty. 
https://cocotb--3717.org.readthedocs.build/en/3717/library_reference.html#cocotb.regression.RegressionManager.register_test